PEP 825: specify how variant markers are evaluated (#73)
wheelnext/peps: peps/pep-0825.rst

-------------------

Four new :ref:`environment markers
<packaging:dependency-specifiers-environment-markers>` are introduced in
dependency specifications. Unlike the markers defined by the
:ref:`packaging:dependency-specifiers` specification, their values are
not the same for every wheel: they are scoped to the variant properties
that the wheel being processed was built for. They MUST be obtained as
described in `evaluating variant markers`_. These are:

1. ``variant_label``: a string, expressing the exact variant label of
   the wheel being processed. For non-variant wheels, it is an empty
   string.
2. ``variant_properties``: a set of all
   ``namespace :: feature :: value`` tuples of the properties

>    corresponding to the variant label that are supported by the target

Reviewer: Good catch. I think we need to fix the implementation here.

```suggestion
   corresponding to the variant label that are compatible with the target
```

I've tried to use "compatible" consistently.

Author: > I think we need to fix the implementation here.

Perhaps you can open an issue for that to not forget?

Reviewer: https://github.com/wheelnext/variantlib/issues/135

   system. For non-variant wheels, it is an empty set.
3. ``variant_features``: a set of all ``namespace :: feature`` pairs
   corresponding to the properties in ``variant_properties``.
4. ``variant_namespaces``: a set of all namespaces of all the properties
   in ``variant_properties``.

The ``variant_label`` marker MUST be matched using the ``==`` and

> ``!=`` operators only. In particular, ``in`` and ``not in`` MUST NOT be

Reviewer: I don't think we should forbid substring matching. The goal of the original statement was that it's a string, while the other markers are sets.

Author: Good call. It looked inconsistent to me, but it wasn't. We just need a reference to https://packaging.python.org/en/latest/specifications/dependency-specifiers/#marker-field-types, which lists the allowed operators. And that will answer questions like "is `variant_label < 'cuda13'` allowed" (it isn't).

I'll update.
